File: src/main/java/io/github/binaryfoo/decoders/bit/NumericBitStringField.java
package io.github.binaryfoo.decoders.bit;

import io.github.binaryfoo.bit.EmvBit;

import java.util.BitSet;
import java.util.Set;

public class NumericBitStringField implements BitStringField {

    private final int byteNumber;
    private final int firstBit;
    private final int lastBit;
    private final String name;

    public NumericBitStringField(int byteNumber, int firstBit, int lastBit, String name) {
        if (lastBit > firstBit) {
            throw new IllegalArgumentException("Must be left to right order: " + lastBit + " > " + firstBit);
        }
        this.byteNumber = byteNumber;
        this.firstBit = firstBit;
        this.lastBit = lastBit;
        this.name = name;
    }

    @Override
    public String getPositionDescription() {
        return String.format("Byte %d Bits %d-%d", byteNumber, firstBit, lastBit);
    }

    @Override
    public String getValueIn(Set<EmvBit> bitstring) {
        BitSet theByte = new BitSet(8);
        for (EmvBit bit : bitstring) {
            if (bit.getByteNumber() == byteNumber && bit.getBitNumber() <= firstBit && bit.getBitNumber() >= lastBit) {
                theByte.set(bit.getBitNumber() - 1, bit.isSet());
            }
        }
        byte[] bytes = theByte.toByteArray();
        int i = bytes.length == 0 ? 0 : (bytes[0] & 0x000000ff) >>> (lastBit - 1);
        return name + " = " + String.valueOf(i);
    }

    @Override
    public int getStartBytesOffset() {
        return byteNumber - 1;
    }

    @Override
    public int getLengthInBytes() {
        return 1;
    }
}

File: src/main/java/io/github/binaryfoo/decoders/bit/NumericBitStringField.kt
package io.github.binaryfoo.decoders.bit

import io.github.binaryfoo.bit.EmvBit

import java.util.BitSet

/**
 * An integer masked out of a number of bits in a single byte.
 */
public class NumericBitStringField(private val byteNumber: Int, private val firstBit: Int, private val lastBit: Int, private val name: String) : BitStringField {

    init {
        if (lastBit > firstBit) {
            throw IllegalArgumentException("Must be left to right order: $lastBit > $firstBit")
        }
    }

    override public fun getPositionIn(bits: Set<EmvBit>?): String {
        return "Byte $byteNumber Bits $firstBit-$lastBit"
    }

    override public fun getValueIn(bits: Set<EmvBit>): String {
        val theByte = BitSet(8)
        for (bit in bits) {
            if (bit.byteNumber == byteNumber && bit.bitNumber <= firstBit && bit.bitNumber >= lastBit) {
                theByte.set(bit.bitNumber - 1, bit.set)
            }
        }
        val bytes = theByte.toByteArray()
        val i = if (bytes.size == 0) 0 else (bytes[0].toInt() and 255).ushr((lastBit - 1))
        return "$name = $i"
    }

    override fun getStartBytesOffset(): Int = byteNumber - 1

    override fun getLengthInBytes(): Int  = 1
}
